### PR TITLE
Fixed#29867 -- cache get_or_set updated to save None as value

### PR DESCRIPTION
**Pull request by ahsanshafiq742**

### CODE REVIEW

File: django/core/cache/backends/memcached.py [hunk 1]
```diff
@@ -2,12 +2,23 @@
 
 import pickle
 import re
+import socket
 import time
 
+import six
```
Comment (auvipy) [inline django/core/cache/backends/memcached.py]: may I know what is the usage of six here?

Comment (ahsanshafiq742) [reply django/core/cache/backends/memcached.py]: @auvipy six dependency removed.

**Comment by ahsanshafiq742** [comment]
@felixxm  This solution is independent of referenced pull requests from memcache (https://github.com/linsomniac/python-memcached/pull/158 and  https://github.com/linsomniac/python-memcached/issues/159).
We can go good without considering the referenced PRs. 
We have successfully handled default value solution including None as value in get method for memcache. Now the get_or_set will behave according to our requirements.

**Comment by ahsanshafiq742** [comment]
@felixxm any updates on it?

**Comment by ahsanshafiq742** [comment]
@felixxm @auvipy 
Do we have any plan for this PR?

File: django/core/cache/backends/memcached.py [hunk 2]
```diff
@@ -160,19 +170,68 @@ def touch(self, key, timeout=DEFAULT_TIMEOUT, version=None):
...
+            return _unsafe_get()
+        except _ConnectionDeadError:
+            # retry once
+            try:
+                if server.connect():
+                    return _unsafe_get()
+                return None
+            except (_ConnectionDeadError, socket.error) as msg:
+                server.mark_dead(msg)
+            return None
```
Comment (felixxm) [inline django/core/cache/backends/memcached.py:224]: I really don't understand why we're introducing a complicated low-level workaround. It looks over engineered. IMO it's not worth to support `None` values on `memcached` if there is no better way to handle them.

@pope1ni Can I ask for your opinion about this patch?

File: django/core/cache/backends/base.py [hunk 3]
```diff
@@ -160,16 +160,15 @@ def get_or_set(self, key, default, timeout=DEFAULT_TIMEOUT, version=None):
 
         Return the value of the key stored or retrieved.
         """
-        val = self.get(key, version=version)
-        if val is None:
+        val = self.get(key, default, version=version)
+        if val in [None, default]:
```
Comment (jacobtylerwalls) [inline django/core/cache/backends/base.py:164]: I'm not understanding why we enter this branch if `val` is `default`. Would you be able to explain? Thanks!

File: django/core/cache/backends/memcached.py (hunk 2 above)
Comment (pope1ni) [reply django/core/cache/backends/memcached.py:224]: @felixxm Will have a look at this tomorrow.

Comment (pope1ni) [reply django/core/cache/backends/memcached.py:224]: @felixxm So the [proposed change](https://github.com/linsomniac/python-memcached/pull/158/files) to support a default argument for get in python-memcached is actually quite simple. But supporting it here in Django requires copying the whole thing, which is not very nice.

Unfortunately the last commit for python-memcached was nearly three years ago now. I've pinged @linsomniac again regarding the possibility of transferring maintainership to Jazzband ([comment](https://github.com/linsomniac/python-memcached/issues/95#issuecomment-725562601)), but if that doesn't proceed I can only suggest that we deprecate support for python-memcached in 3.2 now that we have support for pymemcache as an alternative. If it does move to Jazzband we can fix this issue.

As far as the handling of `None` as a value goes, there are a number of other cache operations that are treating `None` returned from the cache as the key not existing which should also be fixed. I've knocked together a quick thing as a PoC which I can tidy up if you think it looks like a sensible way forward: #13671. We basically just keep the existing behaviour for the python-memcached backend, but fix for all other backends.

Comment (felixxm) [reply django/core/cache/backends/memcached.py:224]: > @felixxm So the [proposed change](https://github.com/linsomniac/python-memcached/pull/158/files) to support a default argument for get in python-memcached is actually quite simple. But supporting it here in Django requires copying the whole thing, which is not very nice.

Yes, it's not.

> Unfortunately the last commit for python-memcached was nearly three years ago now. I've pinged @linsomniac again regarding the possibility of transferring maintainership to Jazzband ([comment](https://github.com/linsomniac/python-memcached/issues/95#issuecomment-725562601)), but if that doesn't proceed I can only suggest that we deprecate support for python-memcached in 3.2 now that we have support for pymemcache as an alternative. If it does move to Jazzband we can fix this issue.

I think it's time to deprecate it, I've created a separate ticket, see ticket-32193.

> As far as the handling of `None` as a value goes, there are a number of other cache operations that are treating `None` returned from the cache as the key not existing which should also be fixed. I've knocked together a quick thing as a PoC which I can tidy up if you think it looks like a sensible way forward: #13671. We basically just keep the existing behaviour for the python-memcached backend, but fix for all other backends.

Agreed, maybe with a small warning in docs.

File: django/core/cache/backends/base.py [hunk 4]
```diff
@@ -160,16 +160,15 @@ def get_or_set(self, key, default, timeout=DEFAULT_TIMEOUT, version=None):
 
         Return the value of the key stored or retrieved.
         """
-        val = self.get(key, version=version)
-        if val is None:
+        val = self.get(key, default, version=version)
```
Comment (jdufresne) [inline django/core/cache/backends/base.py:163]: Did you consider using an opaque value here?


```py
UNSET = object()

...

def get_or_set(self, key, default, timeout=DEFAULT_TIMEOUT, version=None):
        val = self.get(key, UNSET, version=version)
        if val is UNSET:
            ...   # handle set case
```

Doing so would mean a `None` isn't treated special.

Comment (pope1ni) [reply django/core/cache/backends/base.py:163]: @jdufresne See my [comment](https://github.com/django/django/pull/11812#discussion_r522077756) above and the linked PoC in #13671 that does just what you've suggested.

I'll try to polish it soon. We're also going to deprecate python-memcached as it is unmaintained and doesn't support providing a default.

Comment (jdufresne) [reply django/core/cache/backends/base.py:163]: :+1: Nice. Exactly what I had in mind.

**Comment by pope1ni** [comment]
Superseded by #13671.

**Pull request closed without merging**